lib/Dialect/TritonNvidiaGPU/IR/Ops.cpp

```diff
@@ -67,6 +67,27 @@ LogicalResult DotWaitOp::inferReturnTypes(
   return mlir::success();
 }
 
+// -- InitBarrierOp --
+LogicalResult InitBarrierOp::verify() {
+  if (!getAlloc().getType().getElementType().isInteger(64))
+    return emitOpError("barrier allocation must be of i64 type");
+  return success();
+}
+
+// -- WaitBarrierOp --
+LogicalResult WaitBarrierOp::verify() {
+  if (!getAlloc().getType().getElementType().isInteger(64))
+    return emitOpError("barrier allocation must be of i64 type");
+  return success();
+}
+
+// -- AsyncTMACopyGlobalToLocalOp --
+LogicalResult AsyncTMACopyGlobalToLocalOp::verify() {
+  if (!getBarrier().getType().getElementType().isInteger(64))
+    return emitOpError("barrier allocation must be of i64 type");
+  return success();
+}
```

Comment:
Are there no additional invariants associated with AsyncTMACopyGlobalToLocal?

I would have expected constraints on the number of coordinates, on the types of the two memdescs, a constraint on the type of desc_ptr...

(This is a high priority to me because in the past I have multiple times burned multiple days because I was accidentally generating bad IR.)

Reply:
I don't think we can verify any of those statically. There are dependencies between the descriptor itself and the result and coordinate but the descriptor is not statically known. 
The only one would be number of coordinates and rank of the result but even that is not something that has to be the case based on how we are using it.

Let me know if you see anything.

Reply:
Here are some specific constraints that it looks like we could enforce.

- desc_ptr can be literally any pointer?  IIRC a TMA descriptor in PTX is a b64, right?  At least, it has a specific size.  Is it legal to pass desc_ptr as a pointer to a tensor, or a pointer to an i8?  (Will the TMA instruction accept an unaligned i8*?)
- Can `coord` have literally any length?  According to https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#data-movement-and-conversion-instructions-cp-async-bulk-tensor it looks like it must be 1d-5d?
- Can `result` be an arbitrary memdesc?
- (Like in the others, presumably `barrier` needs to be a 1xi64.)

Reply:
> desc_ptr can be literally any pointer? IIRC a TMA descriptor in PTX is a b64, right? At least, it has a specific size. Is it legal to pass desc_ptr as a pointer to a tensor, or a pointer to an i8? (Will the TMA instruction accept an unaligned i8*?)

In PTX pointers aren't typed so yes any pointer works. We could restrict to i8* but I'm not sure it matters. There is some alignment requirements but this is unrelated.

> Can coord have literally any length? According to https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#data-movement-and-conversion-instructions-cp-async-bulk-tensor it looks like it must be 1d-5d?

Right, restricting to <= 5 coordinates make sense. 

> Can result be an arbitrary memdesc?

Not arbitrary but it depends on the descriptor itself as I was saying

Reply:
> We could restrict to i8* but I'm not sure it matters. There is some alignment requirements but this is unrelated.

The only way we have to restrict the alignment of a pointer in Triton is via its type, so type and alignment are related in this way, right?  If you don't want to restrict it in the type system, please add a comment on the alignment requirement if you haven't already.

> Right, restricting to <= 5 coordinates make sense.

Specifically 1 <= coords <= 5.

Reply:
> The only way we have to restrict the alignment of a pointer in Triton is via its type, so type and alignment are related in this way, right? If you don't want to restrict it in the type system, please add a comment on the alignment requirement if you haven't already.

actually the alignment restriction I was thinking about is the alignment of base pointer within the descriptor. I can't find the alignment restrictions on the descriptor in the spec right now.

Reply:
Oh interesting.  OK then!

Reply:
yeah I'm sure there is some restriction but I'll need to reverse engineer it most likely